### Loading Datasets

In [ ]:
import pandas as pd

df = pd.read_csv("/content/heart.csv")

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             918 non-null    int64  
 1   Sex             918 non-null    object 
 2   ChestPainType   918 non-null    object 
 3   RestingBP       918 non-null    int64  
 4   Cholesterol     918 non-null    int64  
 5   FastingBS       918 non-null    int64  
 6   RestingECG      918 non-null    object 
 7   MaxHR           918 non-null    int64  
 8   ExerciseAngina  918 non-null    object 
 9   Oldpeak         918 non-null    float64
 10  ST_Slope        918 non-null    object 
 11  HeartDisease    918 non-null    int64  
dtypes: float64(1), int64(6), object(5)
memory usage: 86.2+ KB


There is a 918 row sample and 12 total columns. <br>
0 Non-Null because all data concistent at 918

__Try to describe data__

In [ ]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Age,918.0,53.510893,9.432617,28.0,47.00,54.0,60.0,77.0
RestingBP,918.0,132.396514,18.514154,0.0,120.00,130.0,140.0,200.0
Cholesterol,918.0,198.799564,109.384145,0.0,173.25,223.0,267.0,603.0
FastingBS,918.0,0.233115,0.423046,0.0,0.00,0.0,0.0,1.0
MaxHR,918.0,136.809368,25.460334,60.0,120.00,138.0,156.0,202.0
Oldpeak,918.0,0.887364,1.066570,-2.6,0.00,0.6,1.5,6.2
HeartDisease,918.0,0.553377,0.497414,0.0,0.00,1.0,1.0,1.0


I think there is a no anomalies and outlier data

__Lets see the preview of dataset__

In [ ]:
df.head()

,Age,Sex,ChestPainType,RestingBP,Cholesterol,FastingBS,RestingECG,MaxHR,ExerciseAngina,Oldpeak,ST_Slope,HeartDisease
0,40,M,ATA,140,289,0,Normal,172,N,0.0,Up,0
1,49,F,NAP,160,180,0,Normal,156,N,1.0,Flat,1
2,37,M,ATA,130,283,0,ST,98,N,0.0,Up,0
3,48,F,ASY,138,214,0,Normal,108,Y,1.5,Flat,1
4,54,M,NAP,150,195,0,Normal,122,N,0.0,Up,0


As we see, there is a 5 categorical columns <br>
There is no columns to drop because all data i think is suitable to target "HeartDisease"

I know target is from read documentation and make sure after see info, describe and preview

__To makesure again i re-check duplicated and missing values__

In [ ]:
print(f"Total duplicated: {df.duplicated().sum()}")

df.isna().sum()

Total duplicated: 0


,0
Age,0
Sex,0
ChestPainType,0
RestingBP,0
Cholesterol,0
FastingBS,0
RestingECG,0
MaxHR,0
ExerciseAngina,0
Oldpeak,0


Yeah there is no duplicated and missing values, so we no nedeed to drop or fill null values

__Sperate numeric and categorical column names__

In [ ]:
num_cols = df.select_dtypes("number").columns
cat_cols = df.select_dtypes("object").columns

print("=== Numberik Columns ===")
print(num_cols)
print("=== Categorical Columns ===")
print(cat_cols)

=== Numberik Columns ===
Index(['Age', 'RestingBP', 'Cholesterol', 'FastingBS', 'MaxHR', 'Oldpeak',
       'HeartDisease'],
      dtype='object')
=== Categorical Columns ===
Index(['Sex', 'ChestPainType', 'RestingECG', 'ExerciseAngina', 'ST_Slope'], dtype='object')


I intentionally display the number of unique values ​​and data types to make analysis easier.

In [ ]:
pd.concat([df.nunique(), df.dtypes], axis=1)

,0,1
Age,50,int64
Sex,2,object
ChestPainType,4,object
RestingBP,67,int64
Cholesterol,222,int64
FastingBS,2,int64
RestingECG,3,object
MaxHR,119,int64
ExerciseAngina,2,object
Oldpeak,53,float64


Check unique values and make sure there is no anomalies values

In [ ]:
for col_name in cat_cols:
  print(f"{col_name} : {df[col_name].unique()}")

Sex : ['M' 'F']
ChestPainType : ['ATA' 'NAP' 'ASY' 'TA']
RestingECG : ['Normal' 'ST' 'LVH']
ExerciseAngina : ['N' 'Y']
ST_Slope : ['Up' 'Flat' 'Down']


There is nothing anomaly values! <br>
Lets make categorical values to numerical

In [ ]:
list_replace = {
    "Sex": {"M": 0, "F": 1},
    "ChestPainType": {'ATA':0, 'NAP': 1, 'ASY':2, 'TA':3},
    "RestingECG": {'Normal':0, 'ST':1, 'LVH':2},
    "ExerciseAngina": {"N":0, "Y":1},
    "ST_Slope": {'Up':0, 'Flat': 1, 'Down': 2}
}

clean_df = df.replace(list_replace)

<ipython-input-22-1f1ed2c38c8d>:9: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  clean_df = df.replace(list_replace)


Lets see preview after replace categorical values to numerical

In [ ]:
clean_df.head()

,Age,Sex,ChestPainType,RestingBP,Cholesterol,FastingBS,RestingECG,MaxHR,ExerciseAngina,Oldpeak,ST_Slope,HeartDisease
0,40,0,0,140,289,0,0,172,0,0.0,0,0
1,49,1,1,160,180,0,0,156,0,1.0,1,1
2,37,0,0,130,283,0,1,98,0,0.0,0,0
3,48,1,2,138,214,0,0,108,1,1.5,1,1
4,54,0,1,150,195,0,0,122,0,0.0,0,0


### Splitting data to training and testing

In [ ]:
from sklearn.model_selection import train_test_split

X = clean_df.drop(columns=['HeartDisease'])
y = clean_df.HeartDisease

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

### Modeling <br>
I use DecisionTreeClassifier because this model is easy to use and suitable with simple database we have

In [ ]:
from sklearn.tree import DecisionTreeClassifier

model = DecisionTreeClassifier(
    max_depth=14,
    criterion="gini",
    splitter="best",
    min_samples_split=2,
    max_features = "sqrt"
)
model.fit(X_train, y_train)
model_pred = model.predict(X_test)

### Scoring <br>
+ __Accuracy__<br>
Measure how often model correct predict / all predictions.<br>
$\frac{TP + TN}{TP + TN + FP + FN}$
+ __Precision__<br>
Measure how often model predict the positive class.<br>
$\frac{TP}{TP + FP}$
+ __Recall__<br>
Measure how often model correct identifies positive instances (TP) from all the actual positive sample, we need FN because to compate how much our model catch the error.<br>
$\frac{TP}{TP + FN}$
+ __F1-Score__<br>
Is combined from precion and recall, the conclusion is combined between how good model to predict correct value and how good model to identifies TP from actual positive sample<br>
$\frac{2 \cdot \text{Precision} \cdot \text{Recal}}{\text{Precision} + \text{Recal}}$

In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

acc_y_pred = accuracy_score(y_test, model_pred)
prec_y_pred = precision_score(y_test, model_pred)
rec_y_pred = recall_score(y_test, model_pred)
f1_y_pred = f1_score(y_test, model_pred)

print(f"Accuracy pred: {acc_y_pred:.3f}")
print(f"Precision pred: {prec_y_pred:.3f}")
print(f"Recall pred: {rec_y_pred:.3f}")
print(f"F1-Score pred: {f1_y_pred:.3f}")

Accuracy pred: 0.821
Precision pred: 0.894
Recall pred: 0.785
F1-Score pred: 0.836
